### PR TITLE
Arm/jvp based selection

### PR DESCRIPTION
**Pull request by AishwwaryaM**
setup for jvp and dynamic k c expts
fix bug to know if weights are distributed to enough trainers before sync aggregation

### CODE REVIEW

File: lib/python/examples/fwdllm/trainer/forward_training/fwdgrad_utils.py [hunk 1]
```diff
@@ -128,6 +128,89 @@ def calculate_var(fwdgrad_list):
...
+    # We calculate variance for each parameter (dim=0), 
+    # then take the mean to get a single scalar representing total noise.
+    actual_var = torch.var(all_grads_stacked, dim=0)
+
+    logger.info(f"shape of actual_var: {actual_var.shape}")
+    logger.info("--- Gradient Distribution Stats ---")
+    logger.info(f"JVP of all updates so far {all_grads_stacked}")
+    log_dist("Signal (Mean)", all_grads_stacked)
+    log_dist("Signal^2", all_grads_stacked**2)
+    # log_dist("Variance", actual_var)
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/trainer/forward_training/fwdgrad_utils.py:202]: We wanted this as well right?

File: lib/python/examples/fwdllm/trainer/forward_training/fwdgrad_utils.py [hunk 2]
```diff
@@ -128,6 +128,89 @@ def calculate_var(fwdgrad_list):
...
+def log_dist(name, tensor):
+    """Helper to log distribution statistics of a tensor."""
+    if tensor.numel() == 0:
+        return
+    
+    # Flatten to ensure we are looking at the distribution of all scalar values
+    flat = tensor.detach().float().view(-1)
+    
+    # Define percentiles to track
+    q = torch.tensor([0.1, 0.25, 0.5, 0.75, 0.9]).to(flat.device)
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/trainer/forward_training/fwdgrad_utils.py]: Nit but can you also include P95 and P99 please

File: lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py [hunk 3]
```diff
@@ -143,7 +179,14 @@ def check_whether_all_receive(self):
     def aggregate(self, current_round):
         start_time = time.time()
         self.var = calculate_var(self.grad_for_var_check_list)
+        var_jvp = calculate_var(self.jvp_for_snr_check_list)
+        self.var_prev_iter_list.append(self.var.item())
+        self.snr = calculate_snr(self.jvp_for_snr_check_list)
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py:189]: I think the functions are modular enough (or can be made to be) to calculate 6 things. For JVP values: {mean^2, var and SNR}. And same for gradient values: {mean^2, var and SNR}. 

Right now I think we are calculating SNR only of JVPs. Verify that all 6 are tracked and printed at each aggregation attempt?

File: lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py [hunk 4]
```diff
@@ -228,6 +271,11 @@ def aggregate(self, current_round):
         if self.args.var_control:
             _force_commit = getattr(self, "_force_commit_this_cycle", False)
             if self.var <= self.var_threshold:
+            # Use different stopping conditions if necessary
+            # if self.var_within_epsilon(): 
+            # if self.snr_within_epsilon_and_var_under(var_jvp):
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py:270]: Slightly confusing about why its commented out. Stopping criteria be enabled via config (e.g. like an ENUM)?

File: lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_99.json [hunk 5]
```diff
@@ -70,6 +70,7 @@
         "learning_rate": 0.01,
         "var_control": true,
         "perturbation_sampling": true,
+        "jvp_perturbation_selection": false,
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_99.json]: rename property to `select_perturbation_using_jvp` as it is more self explanatory?

**Review by dhruvsgarg — COMMENTED** [review]
Important and good PR. Mostly there, need a few nit changes

File: lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py (hunk 4 above)
Comment (AishwwaryaM) [reply lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py:270]: leaving this as is for now

File: lib/python/examples/fwdllm/trainer/forward_training/fwdgrad_utils.py (hunk 1 above)
Comment (AishwwaryaM) [reply lib/python/examples/fwdllm/trainer/forward_training/fwdgrad_utils.py:202]: this is only one value, so, there is no variance

File: lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py (hunk 4 above)
Comment (AishwwaryaM) [reply lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py:270]: as we are not trying out stopping condition much, i will leave this for now

File: lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py (hunk 3 above)
Comment (AishwwaryaM) [reply lib/python/examples/fwdllm/aggregator/FedSgdAggregator.py:189]: not modular, but all 6 are logged

**Review by dhruvsgarg — APPROVED** [review]
LGTM
